doc/src/explanation/best-practices.md

```diff
@@ -0,0 +1,1329 @@
+  second case, we immediately get an error from the undefined name `sine`, and
+  fixing our typo, we see that `expand_trig` can indeed do what we want.
+
+- The biggest gotcha when using string inputs comes from using assumptions. In
+  SymPy, if two symbols have the same name but different assumptions, they are
+  considered unequal:
+
+  ```
+  >>> z1 = symbols('z')
+  >>> z2 = symbols('z', positive=True)
+  >>> z1 == z2
+  False
+  >>> z1 + z2
+  z + z
+  ```
+
+  It is generally recommended to avoid doing this, as it can lead to confusing
+  expressions like the one above (see [](best-practices-defining-symbols)
+  abovee).
+
+  However, strings always parse symbols into symbols without assumptions. So
+  if you have a symbol with an assumption and later try to use the string
+  version of it, you will end up with confusing results.
+
+  ```py
+  >>> from sympy import diff
+  >>> z = symbols('z', positive=True)
+  >>> diff('z**2', z)
+  0
+  ```
+
+  The answer here is apparently wrong, but what happened is that the `z` in
+  `"z**2"` parsed to `Symbol('z')` with no assumptions, which SymPy considers
+  to be a different symbol from `Symbol('z', positive=True)`. So as far as
+  `diff` is concerned, the expression is constant and the result is 0.
+
+  This sort of thing is particularly bad because it generally doesn't lead to
+  any errors. It will just silently give the "wrong" answer because SymPy will
+  be treating symbols that you thought were the same as different. The
+  situation is avoided by not using string inputs.
+
+  If you are parsing strings, and you want some of the symbols in it to have
+  certain assumptions, you should create those symbols and pass them to the
+  dictionary to [`parse_expr()`](sympy.parsing.sympy_parser.parse_expr). For example:
+
+  **Don't**
+
+  ```py
+  >>> a, b, c = symbols('a b c', real=True)
+  >>> # a, b, and c in expr are different symbols without assumptions
+  >>> expr = parse_expr('a**2 + b - c')
+  >>> expr.subs({a: 1, b: 1, c: 1}) # The substitution (apparently) doesn't work
+  a**2 + b - c
+  ```
+
+  **Do**
+
+  ```
+  >>> # a, b, and c are the same as the a, b, c with real=True defined above
+  >>> expr = parse_expr('a**2 + b - c', {'a': a, 'b': b, 'c': c})
+  >>> expr.subs({a: 1, b: 1, c: 1})
+  1
+  ```
+
+- Many operations on SymPy objects work as methods, not functions. These
+  methods won't work on strings, since they are not yet SymPy objects. For
+  example:
+
+  ```
+  >>> "x + 1".subs("x", "y")
+  Traceback (most recent call last):
+  ...
+  AttributeError: 'str' object has no attribute 'subs'
+  ```
+
+  vs.
+
+  ```
+  >>> x, y = symbols('x y')
+  >>> (x + 1).subs(x, y)
+  y + 1
+  ```
+
+- Symbol names can contain any character, including things that aren't
+  valid Python. But if you use strings as input, it is impossible to use such
+  symbols. For example
+
+  ```
+  >>> from sympy import solve
+  >>> solve('x_{2} - 1') # doctest: +SKIP
```

Comment:
I'm not sure if the LaTeX subscript is the best way to choose between `x_2, x_{2}, x2, ...` though
There is a problem that different expressions look same with same subscript, but compares differently `x_2, x_{2}, x2, ...`

Reply:
I agree that this is somewhat contrived. The point here is just to show how things can break, not to give a recommendation on how to name symbols. Anyway, I have seen users making symbol names like this.